### PR TITLE
runtime: Add getDouble functionality to snapshot

### PR DESCRIPTION
**Pull request by tonya11en**
This patch adds the `getDouble` function to the runtime snapshot. This allows for parsing of double values in addition to the existing integer and fractions.

Risk Level: Low
Testing: New unit tests
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by tonya11en** [comment]
@zuercher you make a really good point. I hadn't considered that.

My initial concern with storing numeric values as both `uint64_t` and `double` is that there might be 64-bit int values that are too big to be stored as a double. What about simply swapping the parsing order to parse doubles first and then uints? That way, in your example of code using `Snapshot::getDouble` where the input is changed from 1.5 to 1, we would correctly return 1.0. However, code calling `Snapshot::getInteger` would fail parsing for 1.5 and accurately receive 1.

I'll goa head and make the change I suggested since its trivial, but please let me know your thoughts. Thanks for reviewing.

**Comment by mattklein123** [comment]
> For example, I'm not sure it would be good for getInteger to return 5 for an input of "5.9" but that might be better than having it flip to default values because these they didn't realize a config point only accepted integers vs. floats.

FWIW I think we should probably just floor it and always store as both a double and an integer. I think it will be less surprising. If we decide not to do this can we clearly document that case?

File: source/common/runtime/runtime_impl.h [hunk 1]
```diff
@@ -106,14 +107,21 @@ class SnapshotImpl : public Snapshot,
     if (parseEntryBooleanValue(entry)) {
       return;
     }
-    if (parseEntryUintValue(entry)) {
+
+    if (parseEntryDoubleValue(entry)) {
+      // Valid uint values will always be parseable as doubles, so we assign the value to both the
+      // uint and double fields. In cases where the value is something like "3.1", we will floor the
+      // number by casting it to a uint and assigning the uint value.
+      entry.uint_value_ = entry.double_value_;
```
Comment (zuercher) [inline source/common/runtime/runtime_impl.h]: Let's add a test for negative double values and positive double values >= 2^64. I think we'll want to handle those specially. Probably also need a warning on getInteger that values greater than approximately 2^53 will not be accurately converted to integers.

File: source/common/runtime/runtime_impl.h [hunk 2]
```diff
@@ -106,14 +107,21 @@ class SnapshotImpl : public Snapshot,
...
       return;
     }
+
     parseEntryFractionalPercentValue(entry);
   }
 
   static bool parseEntryBooleanValue(Entry& entry);
   static bool parseEntryUintValue(Entry& entry);
```
Comment (zuercher) [inline source/common/runtime/runtime_impl.h]: Can we remove `parseEntryUintValue`?

File: test/common/runtime/runtime_impl_test.cc [hunk 3]
```diff
@@ -138,6 +138,30 @@ class DiskLoaderImplTest : public LoaderImplTest {
...
+  EXPECT_EQ(1, loader_->snapshot().getInteger("file_with_negative_double", 1));
+  EXPECT_EQ(-4.2, loader_->snapshot().getDouble("file_with_negative_double", 1.1));
+}
+
+TEST_F(DiskLoaderImplTest, DoubleUintLargeIntegerConversion) {
+  setup();
+  run("test/common/runtime/test_data/current", "envoy_override");
+
+  EXPECT_EQ(1, loader_->snapshot().getInteger("file_with_negative_double", 1));
+  EXPECT_EQ(-4.2, loader_->snapshot().getDouble("file_with_negative_double", 1.1));
```
Comment (zuercher) [inline test/common/runtime/runtime_impl_test.cc]: Should this be a different file? Seems like as-is it's the same test as DoubleUintInteractionNegatives.

Comment (tonya11en) [reply test/common/runtime/runtime_impl_test.cc]: Yeah, sorry that push was half-baked. That's why I added the /wait above- didn't want you to waste time reviewing this yet.

**Comment by zuercher** [comment]
Cool. Let's add a comment on getInteger for values > 2^53 and then I think this is good to go.

File: docs/root/intro/version_history.rst [hunk 4]
```diff
@@ -66,6 +66,7 @@ Version history
 * upstream: added new :ref:`failure-percentage based outlier detection<arch_overview_outlier_detection_failure_percentage>` mode.
 * upstream: use p2c to select hosts for least-requests load balancers if all host weights are the same, even in cases where weights are not equal to 1.
 * upstream: added :ref:`fail_traffic_on_panic <envoy_api_field_Cluster.CommonLbConfig.ZoneAwareLbConfig.fail_traffic_on_panic>` to allow failing all requests to a cluster during panic state.
+* runtime: allow for the ability to parse integers as double values and vice-versa.
```
Comment (mattklein123) [inline docs/root/intro/version_history.rst]: nit: alpha order.

**Comment by mattklein123** [comment]
Sorry needs a master merge.

/wait

File: docs/root/intro/version_history.rst [hunk 5]
```diff
@@ -59,6 +59,7 @@ Version history
 * router check tool: add comprehensive coverage reporting.
 * router check tool: add deprecated field check.
 * router check tool: add flag for only printing results of failed tests.
+* runtime: allow for the ability to parse integers as double values and vice-versa.
```
Comment (mattklein123) [inline docs/root/intro/version_history.rst]: Sorry needs another merge fix.

/wait

**Pull request merged**